src/core/update_initial_conditions.jl

```diff
@@ -0,0 +1,326 @@
+) where {T <: PSY.HydroGen}
+    _make_initial_conditions!(
+        container,
+        devices,
+        ICKey(DevicePower, T),
+        _make_initial_condition_active_power,
+        _get_active_power_output_value,
+        # Doesn't require Cache
+    )
+
+    return
+end
+
+function duration_init(
+    container::InitialConditions,
+    devices::IS.FlattenIteratorWrapper{T},
+) where {T <: PSY.HydroGen}
+    for key in (ICKey(TimeDurationON, T), ICKey(TimeDurationOFF, T))
+        _make_initial_conditions!(
+            container,
+            devices,
+            key,
+            _make_initial_condition_active_power,
+            _get_active_power_duration_value,
+            TimeStatusChange,
+        )
+    end
+
+    return
+end
+
+# TODO: This IC needs a cache for Simulation over long periods of time
+function storage_energy_init(
+    container::InitialConditions,
+    devices::IS.FlattenIteratorWrapper{T},
+) where {T <: PSY.HydroGen}
+    key = ICKey(EnergyLevel, T)
+    _make_initial_conditions!(
+        container,
+        devices,
+        key,
+        _make_initial_condition_reservoir_energy,
+        _get_reservoir_energy_value,
+        StoredEnergy,
+    )
+
+    return
+end
+
+function _make_initial_conditions!(
+    container::InitialConditions,
+    devices::IS.FlattenIteratorWrapper{T},
+    key::ICKey,
+    make_ic_func::Function,
+    get_val_func::Function,
+    cache = nothing,
+) where {T <: PSY.Device}
+    length_devices = length(devices)
+
+    if !has_initial_conditions(container, key)
+        @debug "Setting $(key.ic_type) initial conditions for the status of all devices $(T) based on system data"
+        ini_conds = Vector{InitialCondition}(undef, length_devices)
+        set_initial_conditions!(container, key, ini_conds)
+        for (ix, dev) in enumerate(devices)
+            ini_conds[ix] = make_ic_func(container, dev, get_val_func(dev, key), cache)
+        end
+    else
+        ini_conds = get_initial_conditions(container, key)
+        ic_devices = Set((IS.get_uuid(ic.device) for ic in ini_conds))
+        for dev in devices
+            IS.get_uuid(dev) in ic_devices && continue
+            @debug "Setting $(key.ic_type) initial conditions for the status device $(PSY.get_name(dev)) based on system data"
+            push!(ini_conds, make_ic_func(container, dev, get_val_func(dev, key), cache))
+        end
+    end
+
+    @assert length(ini_conds) == length_devices
+    return
+end
+
+function _make_initial_condition_active_power(
+    container,
+    device::T,
+    value,
+    cache = nothing,
+) where {T <: PSY.Component}
+    return InitialCondition(device, _get_ref_active_power(T, container), value, cache)
+end
+
+function _make_initial_condition_energy(
+    container,
+    device::T,
+    value,
+    cache = nothing,
+) where {T <: PSY.Component}
+    return InitialCondition(device, _get_ref_energy(T, container), value, cache)
+end
+
+function _make_initial_condition_reservoir_energy(
```

Comment:
Can these make functions be generalized to use multiple dispatches instead of having each named individually? It is not totally clear to me (right now) what the new flow of actions will be with these changes.

Reply:
No logic in this file actually changed.  It got moved from another file.  I also changed PSIContainer -> InitialConditions.  Rather that try to optimize it, I'd prefer to make the prototype of building initial conditions from a specification that I discussed with you earlier this week.

Reply:
Ok, I guess this is an improvement for later